cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStore.java

```diff
@@ -204,108 +207,109 @@ private void persistJobAndTrigger(final JobDetail newJob, final OperableTrigger
         triggerState = super.getTriggerState(newTrigger.getKey());
       }
       final Trigger.TriggerState finalTriggerState = triggerState;
-      factory.createExecutor(ImmutableList.of((TransactionAware) table))
-        .execute(() -> {
-          if (newJob != null) {
-            persistJob(table, newJob);
-            LOG.debug("Schedule: stored job with key {}", newJob.getKey());
-          }
-          if (newTrigger != null) {
-            persistTrigger(table, newTrigger, finalTriggerState);
-            LOG.debug("Schedule: stored trigger with key {}", newTrigger.getKey());
-          }
-        });
+      TransactionRunners.run(transactionRunner, context -> {
+        StructuredTable table = getTimeScheduleStructuredTable(context);
+        if (newJob != null) {
+          persistJob(table, newJob);
+          LOG.debug("Schedule: stored job with key {}", newJob.getKey());
+        }
+        if (newTrigger != null) {
+          persistTrigger(table, newTrigger, finalTriggerState);
+          LOG.debug("Schedule: stored trigger with key {}", newTrigger.getKey());
+        }
+      });
     } catch (Throwable th) {
       throw Throwables.propagate(th);
     }
   }
 
   // Persist the job information to dataset
-  private void persistJob(Table table, JobDetail job) {
-    byte[][] cols = new byte[1][];
-    byte[][] values = new byte[1][];
+  private void persistJob(StructuredTable table, JobDetail job) throws IOException {
+    upsert(table, JOB_KEY, job.getKey().getName(), SerializationUtils.serialize(job));
+  }
 
-    cols[0] = Bytes.toBytes(job.getKey().toString());
-    values[0] = SerializationUtils.serialize(job);
-    table.put(JOB_KEY, cols, values);
+  private void delete(StructuredTable table, String type, String name) throws IOException {
+    table.delete(getPrimaryKeys(type, name));
   }
 
-  @VisibleForTesting
-  void removeJob(Table table, JobKey key) {
-    byte[][] col = new byte[1][];
-    col[0] = Bytes.toBytes(key.toString());
-    table.delete(JOB_KEY, col);
+
+  @Nullable
+  private byte[] get(StructuredTable table, String type, String name) throws IOException {
+    Optional<StructuredRow> row = table.read(getPrimaryKeys(type, name));
+    if (!row.isPresent()) {
```

Comment:
would this be better `row.map(structuredRow -> structuredRow.getBytes(StoreDefinition.TimeScheduleStore.VALUE_FIELD)).orElse(null);`?

Reply:
imo prefer not to map when we're operating on a singleton